### PR TITLE
[Sealing] Sealing edge case for phase 3 

### PR DESCRIPTION
**Pull request by durkmurder**
[-> dapperlabs/flow-go/issues/5570](https://github.com/dapperlabs/flow-go/issues/5570)

### Context:
In current logic on startup we are traversing all valid descendants to repopulate approvals tree. To calculate SoR target block needs to have a child and we don't satisfy this condition. 

In phase 2 it's not an issue since we take SoR from executed block which has a child by definition. For phase 3 it's not the case.

### What was done
- Extended tests to support both phase 2 and phase 3 behaviour, need to be adjusted when moving into phase 3.
- Changed traversing logic to include results based on parent not actual block, this way we will skip blocks without children

### CODE REVIEW

File: engine/consensus/sealing/core.go [hunk 1]
```diff
@@ -202,12 +202,22 @@ func (c *Core) RepopulateAssignmentCollectorTree(payloads storage.Payloads) erro
...
 	for _, blockID := range validPending {
 		block, err := c.headers.ByBlockID(blockID)
 		if err != nil {
 			return fmt.Errorf("could not retrieve header for unfinalized block %x: %w", blockID, err)
 		}
-		err = resultProcessor(block)
+
+		parent, err := c.headers.ByBlockID(block.ParentID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve header for unfinalized block %x: %w", block.ParentID, err)
```
Comment (zhangchiqing) [inline engine/consensus/sealing/core.go:217]: better to update the error message accordingly

**Review by zhangchiqing — COMMENTED** [review]
Changes look good. But better to merge into a feature branch for phase 3, so that we could launch all phase 3 changes together after phase 2 have been tested

**Review by AlexHentschel — APPROVED** [review]
🎸

**Comment by AlexHentschel** [comment]
I don't think we need to merge this into a feature branch. It can go right to master:
* in the main logic, we already process the parent of incorporated blocks (-> [code](https://github.com/onflow/flow-go/blob/d16730ac4245a923ea7e39e90329949b6ed3b90a/engine/consensus/sealing/engine.go#L449-L454))
* In the old implementation, the re-population algorithm used a _different_ convention. With Yurii's PR we now make it consistent (we got lucky that the inconsistency in the past version didn't cause any problems in Phase 2c). 

I view this PR more as cleaning up tech debt from Phase 2c rather then adding breaking features for Phase 3.

File: engine/consensus/sealing/core.go [hunk 2]
```diff
@@ -203,12 +203,22 @@ func (c *Core) RepopulateAssignmentCollectorTree(payloads storage.Payloads) erro
 
 	c.log.Info().Msgf("reloading assignments from %d unfinalized blocks into collector tree", len(validPending))
 
+	// When adding an incorporated result to the AssignmentCollectorTree, the verifier assignment is computed.
+	// This requires the Source of Randomness (SoR). However, the SoR is only available, if a valid child block exists.
+	// That is why we need to traverse blocks that already have a valid child
+	// This situation cannot happen in phase 2 because SoR is taken from executed block(not the one which incorporates result)
```
Comment (zhangchiqing) [inline engine/consensus/sealing/core.go:209]:
```suggestion
	// We use AssignmentCollectorTree for collecting approvals for each incorporated result. 
	// In order to verify the received approvals, the verifier assignment for each incorporated result 
	// needs to be known.
	// The verifier assignment is random, its Source of Randomness (SoR) is only available if a valid 
	// child block exists.
	// In other words, the parent of a valid block must have the SoR available. Therefore, we traverse
	// through valid pending blocks which already have a valid child, and load each result in those block
	// into the AssignmentCollectorTree.
```

**Review by zhangchiqing — APPROVED** [review]
👍

**Pull request merged**